### PR TITLE
[tune] Fix trial runner/controller whitelist attributes

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

With PBT and BOHB, we currently can get these warnings:

```
2023-05-24 04:03:59,962 WARNING trial_runner.py:1543 -- You are trying to access pause_trial interface of TrialRunner in TrialScheduler, which is being restricted. If you believe it is reasonable for your scheduler to access this TrialRunner API, please reach out to Ray team on GitHub. A more strict API access pattern would be enforced starting 1.12s.0
```

While technically we should have seen those warnings before (in the old execution path), it looks like they only come up ever since we activated the new execution path.

This PR adds the rest of those attributes we access in our tune-provided schedulers to the whitelist to get rid of the warnings. 

### Background

We introduced the warnings as preparation to refactor the scheduler interface. We wanted to discourage users to directly call trial runner/executor APIs. The plan was to redefine the interface and get rid of these calls in our own implementations. However, we have deprioritized this effort - that's why we keep a whitelist. We should revisit fixing the scheduler interface in the future.

## Related issue number

Addresses part of #35640

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/execution/tune_controller.py [hunk 1]
```diff
@@ -154,8 +154,20 @@ def _wrapped(self):
...
+                "get_live_trials",
+                "_set_trial_status",
+                "pause_trial",
+                "stop_trial",
+            },
+            executor_whitelist_attr={
+                "has_resources_for_trial",
+                "pause_trial",
+                "save",
+                "resource_updater",
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:169]:
```suggestion
                "_resource_updater",
```

File: python/ray/tune/execution/tune_controller.py [hunk 2]
```diff
@@ -154,8 +154,20 @@ def _wrapped(self):
         return TrialRunnerWrapper(
             self,
             trial_executor=_FakeRayTrialExecutor(self),
-            runner_whitelist_attr={"search_alg", "get_trials", "_set_trial_status"},
-            executor_whitelist_attr={"has_resources_for_trial", "pause_trial", "save"},
+            runner_whitelist_attr={
+                "search_alg",
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:158]:
```suggestion
                "_search_alg",
```

File: python/ray/tune/execution/tune_controller.py [hunk 3]
```diff
@@ -154,8 +154,20 @@ def _wrapped(self):
...
+                "get_trials",
+                "get_live_trials",
+                "_set_trial_status",
+                "pause_trial",
+                "stop_trial",
+            },
+            executor_whitelist_attr={
+                "has_resources_for_trial",
+                "pause_trial",
+                "save",
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:168]: For these methods, we can do:

```
self.get_trials.__name__
```

This way we don't ever allow/exclude a non-existent method.

File: python/ray/tune/execution/tune_controller.py [hunk 4]
```diff
@@ -154,8 +154,20 @@ def _wrapped(self):
         return TrialRunnerWrapper(
             self,
             trial_executor=_FakeRayTrialExecutor(self),
-            runner_whitelist_attr={"search_alg", "get_trials", "_set_trial_status"},
-            executor_whitelist_attr={"has_resources_for_trial", "pause_trial", "save"},
+            runner_whitelist_attr={
```
Comment (justinvyu) [inline python/ray/tune/execution/tune_controller.py:157]: For the `TuneController`, I think we need to more keys:

```
"_schedule_trial_pause", "_schedule_trial_save", "_resource_updater"
```

The `FakeRayTrialExecutor` just passes through to these methods of the wrapped TuneController, which would still show the warning message.

Comment (justinvyu) [reply python/ray/tune/execution/tune_controller.py:157]: The `executor_whitelist_attr` also needs to whitelist `_tune_controller`.

**Review by justinvyu — COMMENTED** [review]
Have a few suggestions!

File: python/ray/tune/execution/tune_controller.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/tune/execution/tune_controller.py:168]: Or we can also do an assert within the wrapper class to make sure `getattr` returns something for each attr.

File: python/ray/tune/execution/tune_controller.py (hunk 4 above)
Comment (krfricke) [reply python/ray/tune/execution/tune_controller.py:157]: I don't think this is the case - the `FakeRayTrialExecutor` wraps around the actual TuneController, so it doesn't trigger any warning.

We also don't want to expose the tune controller attribute at all (the "real" executor won't have that property so we wouldn't want to use it anyway)

File: python/ray/tune/execution/tune_controller.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/execution/tune_controller.py:158]: These are attributes that the developer of the schedulers can use - they shouldn't access private properties

File: python/ray/tune/execution/tune_controller.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/execution/tune_controller.py:169]: see above

File: python/ray/tune/execution/tune_controller.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/tune/execution/tune_controller.py:158]: Oh, I didn't see that `search_alg` was a property defined in the base tune controller, makes sense!

File: python/ray/tune/execution/tune_controller.py (hunk 4 above)
Comment (justinvyu) [reply python/ray/tune/execution/tune_controller.py:157]: Ah got it, my mistake!

File: python/ray/tune/execution/tune_controller.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/tune/execution/tune_controller.py:169]: I think this one actually doesn't have a public counterpart.

**Review by justinvyu — COMMENTED** [review]
Thanks! Just need to fix the `resource_updater` attribute, the `hasattr` check seems to be causing all the tests to fail.

**Review by justinvyu — APPROVED** [review]

**Pull request merged**